bazarr/subtitles/upgrade.py

```diff
@@ -276,16 +287,47 @@ def get_upgradable_episode_subtitles():
                                       TableHistory.score.is_not(None),
                                       (TableHistory.score < 357)]
     upgradable_episodes_conditions += get_exclusion_clause('series')
-    return select(TableHistory.id)\
-        .select_from(TableHistory) \
+    subtitles_to_upgrade = database.execute(
+        select(TableHistory.id,
+               TableHistory.video_path,
+               TableHistory.language,
+               TableHistory.upgradedFromId)
+        .select_from(TableHistory)
+        .join(TableShows, onclause=TableHistory.sonarrSeriesId == TableShows.sonarrSeriesId)
+        .join(TableEpisodes, onclause=TableHistory.sonarrEpisodeId == TableEpisodes.sonarrEpisodeId)
         .join(max_id_timestamp, onclause=and_(TableHistory.video_path == max_id_timestamp.c.video_path,
                                               TableHistory.language == max_id_timestamp.c.language,
-                                              max_id_timestamp.c.timestamp == TableHistory.timestamp)) \
-        .join(TableShows, onclause=TableHistory.sonarrSeriesId == TableShows.sonarrSeriesId) \
-        .join(TableEpisodes, onclause=TableHistory.sonarrEpisodeId == TableEpisodes.sonarrEpisodeId) \
-        .where(reduce(operator.and_, upgradable_episodes_conditions)) \
-        .order_by(TableHistory.timestamp.desc())\
-        .subquery()
+                                              max_id_timestamp.c.timestamp == TableHistory.timestamp))
+        .where(reduce(operator.and_, upgradable_episodes_conditions))
+        .order_by(TableHistory.timestamp.desc())) \
+        .all()
+
+    query_actions_without_upgrade = [x for x in query_actions if x != 3]
+    upgradable_episode_subtitles = {}
+    for subtitle_to_upgrade in subtitles_to_upgrade:
+        # check if we have the original subtitles id in database and use it instead of guessing
+        if subtitle_to_upgrade.upgradedFromId:
+            upgradable_episode_subtitles.update({subtitle_to_upgrade.id: subtitle_to_upgrade.upgradedFromId})
+            continue
+
+        # if not, we have to try to guess the original subtitles id
+        potential_parents = database.execute(
+            select(TableHistory.id, TableHistory.action)
+            .where(TableHistory.video_path == subtitle_to_upgrade.video_path,
+                   TableHistory.language == subtitle_to_upgrade.language,)
+            .order_by(TableHistory.timestamp.desc())
+        ).all()
+
+        confirmed_parent = None
+        for potential_parent in potential_parents:
+            if potential_parent.action in query_actions_without_upgrade:
+                confirmed_parent = potential_parent.id
+                break
+
+        if confirmed_parent not in upgradable_episode_subtitles.values():
+            upgradable_episode_subtitles.update({subtitle_to_upgrade.id: confirmed_parent})
+
+    return upgradable_episode_subtitles
```

Comment:
There is so many combinations and possibilities here, i would say that adding couple of debug logs won't hurt, what do you think?